### ISSUE TITLE
Indent Guides

### ISSUE THREAD

**Issue by LunarLambda** [neovim/neovim#9337]
## Preface

Yes, there are plugins for that. They are quite far from perfect, however.
I know it's a fairly complex feature, but please bear with me for a moment.

## Rationale

Every major programming-oriented text editor supports indent guides. Sublime, Atom, VS Code, even the rather innocuous Notepad++. If you look for 'Vim indentation guides' online you'll find _loads_ of posts about it. People want this feature.

## How big is 'one indent'?

Vim supports many types of indentation. This makes it somewhat difficult to determine how big 'one indent' should be. We can disregard tabs versus spaces here since what matters is only the size of the indentation.

Consider how the following options are used:
```
Option           Use
--------------------
shiftwidth       cindent, >>, <<, etc, and <Tab> when smarttab or softtabstop < 0
tabstop          <Tab> if nosmarttab and softtabstop = 0
softtabstop      <Tab> if nosmarttab and softtabstop > 0
```

The size of indentation inserted when pressing <Tab> can be expressed with the following pseudocode then:

```c
#include <stdint.h> // intptr_t

// Returns 0/1 for boolean options, integers for numeric options, const char * (as intptr_t) for string options
intptr_t get_option(const char *option);

// :help shiftwidth()
int shiftwidth();

int get_indent_size()
{
    int sts = (int)get_option("softtabstop");

    if (get_option("smarttab") || sts < 0)
    {
        return shiftwidth();
    }
    else if (sts > 0)
    {
        return sts;
    }
    else
    {
        return (int)get_option("tabstop");
    }
}
```

## What else is needed?

-  An 'indent' value for 'listchars' to change how they're displayed (default: ???)
- Or: An 'indentguides' option (single char or empty (disabled), default: "│" or "|")

It would also help if indentation became a 'first-class concept' as discussed here https://github.com/neovim/neovim/issues/6052#issuecomment-277306274

## Costs

- Implementation: high (might be a lot of UI code?)
- Maintainability: low (unless new indentation options are added)

But: Huge quality of life feature!

## A look into the future?

`:set tabstop=4 shiftwidth=4 softtabstop=0 noexpandtab listchars=tab:│\ list`
![image](https://user-images.githubusercontent.com/38919842/49704034-c3875200-fc0d-11e8-917a-29227efa8d0d.png)

Unfortunately, this only works if you're indenting with hard tabs only, and using tabs for nothing else.

**Comment by and3rson** [comment]
Is this still being worked on? Would totally love to see this!

**Comment by theHamsta** [comment]
You're looking for something like this: https://github.com/glepnir/indent-guides.nvim . It requires the lastest and greatest master since nvim can now display virtual text everywhere on normal text.

**Comment by mfiano** [comment]
I'm also looking forward to a proper solution like other editors that is smart about the first non-whitespace character of its starting block, rather than just highlighting odd/even columns

**Comment by theHamsta** [comment]
@mfiano I don't really get what you mean. The plugin respects your indent settings (each highlight is one indent unit whether this is one tab, two spaces or four spaces) and it only highlights character until the first non-whitespace.

**Comment by zegervdv** [comment]
have you tried: https://github.com/lukas-reineke/indent-blankline.nvim ? 
It does full indents now, not just the blank lines if you use the lua branch.

**Comment by bfredl** [comment]
closed as there are working plugins (for 0.5)

**Comment by WhyNotHugo** [comment]
https://github.com/lukas-reineke/indent-blankline.nvim's `lua` branch works like a charm. Be sure to check out the help page, the README doesn't mention most of the options.
